maas/resource_maas_boot_source_selection.go

```diff
@@ -0,0 +1,227 @@
+		return diag.FromErr(err)
+	}
+
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	bootsourceselection, err := getBootSourceSelection(client, bootsource.ID, id)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	bootsourceselectionParams := entity.BootSourceSelectionParams{
+		OS:        d.Get("os").(string),
+		Release:   d.Get("release").(string),
+		Arches:    convertToStringSlice(d.Get("arches").(*schema.Set).List()),
+		Subarches: convertToStringSlice(d.Get("subarches").(*schema.Set).List()),
+		Labels:    convertToStringSlice(d.Get("labels").(*schema.Set).List()),
+	}
+
+	if _, err := client.BootSourceSelection.Update(bootsource.ID, bootsourceselection.ID, &bootsourceselectionParams); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return resourceBootSourceSelectionRead(ctx, d, meta)
+}
+
+func resourceBootSourceSelectionDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ClientConfig).Client
+
+	bootsource, err := getBootSource(client)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	bootsourceselection, err := getBootSourceSelection(client, bootsource.ID, id)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	// if the boot source selection is the default, we need to treat it differently
+	defaultbootsourceselection, err := fetchDefaultBootSourceSelection(client)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	if defaultbootsourceselection.BootSourceID == bootsourceselection.BootSourceID && defaultbootsourceselection.ID == bootsourceselection.ID {
+		bootsourceselectionParams := entity.BootSourceSelectionParams{
+			OS:        defaultbootsourceselection.OS,
+			Release:   defaultbootsourceselection.Release,
+			Arches:    []string{"amd64"},
+			Subarches: []string{"*"},
+			Labels:    []string{"*"},
+		}
+		if _, err := client.BootSourceSelection.Update(bootsource.ID, bootsourceselection.ID, &bootsourceselectionParams); err != nil {
+			return diag.FromErr(err)
+		}
+	} else {
+		return diag.FromErr(client.BootSourceSelection.Delete(bootsource.ID, bootsourceselection.ID))
+	}
+
+	return nil
+}
+
+func fetchDefaultBootSourceSelection(client *client.Client) (*entity.BootSourceSelection, error) {
+	// Fetch the default commissioning details
+	var default_os string
+
+	default_os_bytes, err := client.MAASServer.Get("default_osystem")
+	if err != nil {
+		return nil, err
+	}
+
+	err = json.Unmarshal(default_os_bytes, &default_os)
+	if err != nil {
+		return nil, err
+	}
+
+	var default_release string
+
+	default_release_bytes, err := client.MAASServer.Get("default_distro_series")
+	if err != nil {
+		return nil, err
+	}
+
+	err = json.Unmarshal(default_release_bytes, &default_release)
+	if err != nil {
+		return nil, err
+	}
+
+	// Only then fetch the default bootsource that refers to them
+
+	bootsource, err := getBootSource(client)
```

Comment:
This is not needed. Actually, the check should be on the `d.os == "ubuntu" && d.release == commissioning_distro_series`. There is no need to fetch the default selection from MAAS.